### PR TITLE
Core: Restore `/preview` etc package exports; return unresolved path from presets.

### PR DESCRIPTION
**Pull request by tmeasday**
Issue:

Chromatic regressions introduced in https://github.com/storybookjs/storybook/pull/19007, ultimately due to many addons not working in example apps.

The root issue that was being worked around was that if a package has an `exports` field, Vite can only import an entry there. i.e. it can `import('@storybook/addon-backgrounds/preview')` not `import('@storybook/addon-backgrounds/dist/preview.js')`, which is what the presets code was resolving it to.

## What I did

In the presets code, try resolving an import path, but return the import path unmodified, so the builder can import it itself.

NOTE: This means that import paths (e.g. `previewEntries`) will be of the form `'@storybook/addon-backgrounds/preview'` rather than `/Users/xyz/..../node_modules/@storybook/addon-backgrounds/preview`. Is this a problem @ndelangen? If so we could "half resolve" the import (ie. absolutize the import path).

## How to test

- Run example app (in both Vite and Webpack), check the addons, a11y, etc addons are working.
- Chromatic changes should be gone on this PR.

### CODE REVIEW

File: code/lib/core-common/src/presets.ts [hunk 1]
```diff
@@ -93,10 +93,10 @@ export const resolveAddonName = (
   const path = name;
 
   // when user provides full path, we don't need to do anything!
-  const managerFile = r(`${path}/manager`);
+  const managerFile = r(`${path}/manager`) || r(`${path}/manager.js`);
```
Comment (yannbf) [inline code/lib/core-common/src/presets.ts:96]: Is it possible to try and check whether such file exists, regardless of extension, then get its full path back, with the extension? like we did for `requireResolve`. People not only do `.js`, they can do `.mjs` and maybe other extensions too

**Review by shilman — APPROVED** [review]
LGTM. @ndelangen we should probably figure out how to clean this up somehow -- maybe in Vite, which was the reason for the change in the first place?

File: code/lib/core-common/src/presets.ts (hunk 1 above)
Comment (kylegach) [reply code/lib/core-common/src/presets.ts:96]: Team decided this will be addressed later.

**Comment by IanVS** [comment]
The reason vite is importing the files with `.js` is that's what we get from `await presets.apply('config', [], options)`;

File: code/lib/core-common/src/presets.ts (hunk 1 above)
Comment (tmeasday) [reply code/lib/core-common/src/presets.ts]: We looked at this but ended up doing the opposite. We just check `/manager` now, all systems (node, vite and wp) can resolve that fine.

**Comment by tmeasday** [comment]
@ndelangen -- @shilman and I reworked this, and we are happy with this solution now, it resolves (haha) everything, apart from one possible wrinkle (see above):

> NOTE: This means that import paths (e.g. previewEntries) will be of the form '@storybook/addon-backgrounds/preview' rather than /Users/xyz/..../node_modules/@storybook/addon-backgrounds/preview. Is this a problem @ndelangen? If so we could "half resolve" the import (ie. absolutize the import path).

**Review by ndelangen — APPROVED** [review]
bingo! this is great!

when I opened this PR this morning I instantly had the realization that this was the problem, and then 1 click later, you've implemented this exactly like that.

I guess I really needed sleep yesterday.

Awesome work @tmeasday @shilman

File: code/lib/core-common/src/presets.ts (hunk 1 above)
Comment (yannbf) [reply code/lib/core-common/src/presets.ts]: Awesome!

File: code/lib/core-common/src/presets.ts [hunk 2]
```diff
@@ -72,6 +72,7 @@ export const resolveAddonName = (
   options: any
 ): ResolvedAddonPreset | ResolvedAddonVirtual | undefined => {
   const r = name.startsWith('/') ? safeResolve : safeResolveFrom.bind(null, configDir);
+  const ifResolves = (path: string) => r(path) && path;
```
Comment (yannbf) [inline code/lib/core-common/src/presets.ts:75]: What about renaming to getResolvedPath?

Comment (IanVS) [reply code/lib/core-common/src/presets.ts:75]: If I'm understanding correctly, this doesn't return the resolved path, it checks if the resolved path is truthy, and then returns the original path, or `undefined`.

**Review by IanVS — APPROVED** [review]
This is great, and means I think we can get rid of our [custom utility](https://github.com/storybookjs/storybook/blob/next/code/lib/builder-vite/src/utils/transform-abs-path.ts) that converted absolute paths back into "bare" imports.

File: code/lib/core-common/src/presets.ts (hunk 2 above)
Comment (yannbf) [reply code/lib/core-common/src/presets.ts:75]: Ah I see. I read it quickly and didn't notice that 🤦 thanks!

**Comment by tmeasday** [comment]
> This is great, and means I think we can get rid of our [custom utility](https://github.com/storybookjs/storybook/blob/next/code/lib/builder-vite/src/utils/transform-abs-path.ts?rgh-link-date=2022-08-30T11%3A17%3A34Z) that converted absolute paths back into "bare" imports.

As discussed, hold your horses on this, absolute paths are coming back!

> If I'm understanding correctly, this doesn't return the resolved path, it checks if the resolved path is truthy, and then returns the original path, or undefined.

That's correct. But I am going to tweak it to absolutize the path (but not resolve the export).

**Comment by ndelangen** [comment]
This is weird... so sb-bench IS running on localhost, but not reachable (assumed)?

**Comment by ndelangen** [comment]
argh, seems it's not consistent @tmeasday nothing changed between the CI being green and showing the error again.

What a pain.

**Pull request merged**